### PR TITLE
enhance: support iterative filter execution

### PR DESCRIPTION
**Pull request by chasingegg**
issue: #37360

### CODE REVIEW

File: internal/core/src/exec/QueryContext.h [hunk 1]
```diff
@@ -230,6 +230,17 @@ class QueryContext : public Context {
...
 
+    knowhere::MetricType
+    get_metric_type() {
+        return search_info_.metric_type_;
+    }
+
+    std::optional<int64_t>
+    get_iterator_batch_size() {
+        return milvus::index::GetValueFromConfig<int64_t>(
+            search_info_.search_params_, "batch_size");
```
Comment (PwzXxm) [inline internal/core/src/exec/QueryContext.h:241]: Is it possible to rename it as conflicting to E2E Iterator parameters?

File: internal/core/src/exec/expression/Expr.h [hunk 2]
```diff
@@ -204,13 +220,16 @@ class SegmentExpr : public Expr {
 
     void
     MoveCursor() override {
-        if (is_index_mode_) {
-            MoveCursorForIndex();
-            if (segment_->HasFieldData(field_id_)) {
+        // when we specify input, do not maintain states
+        if (has_offset_input_) {
```
Comment (PwzXxm) [inline internal/core/src/exec/expression/Expr.h:224]: Why it does not need to move cursor here when input is specified? Not the other way around as in `internal/core/src/exec/expression/LogicalBinaryExpr.h` per say?

File: internal/core/src/exec/operator/FilterNode.cpp [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+}
+
+template <bool large_is_better>
+inline size_t
+find_binsert_position(const std::vector<float>& distances,
+                      size_t lo,
+                      size_t hi,
+                      float dist) {
+    while (lo < hi) {
+        size_t mid = (lo + hi) >> 1;
```
Comment (PwzXxm) [inline internal/core/src/exec/operator/FilterNode.cpp:56]: Potential overflow, use `size_t mid = lo + ((hi - lo) >> 1)`

File: internal/core/src/exec/operator/FilterNode.cpp [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+            nq_index++;
+        }
+    }
+    query_context_->set_search_result(std::move(search_result));
+    std::chrono::high_resolution_clock::time_point scalar_end =
+        std::chrono::high_resolution_clock::now();
+    double scalar_cost =
+        std::chrono::duration<double, std::micro>(scalar_end - scalar_start)
+            .count();
+    monitor::internal_core_search_latency_postfilter.Observe(scalar_cost);
```
Comment (PwzXxm) [inline internal/core/src/exec/operator/FilterNode.cpp:186]: Great to see adding metrics as well!

File: internal/core/src/exec/operator/FilterNode.cpp [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+                    results_.size() == 1 && results_[0] != nullptr,
+                    "PhyFilterNode result size should be size one and not "
+                    "be nullptr");
+
+                auto col_vec =
+                    std::dynamic_pointer_cast<ColumnVector>(results_[0]);
+                auto col_vec_size = col_vec->size();
+                TargetBitmapView bitsetview(col_vec->GetRawData(),
+                                            col_vec_size);
+                Assert(bitsetview.size() <= batch_size);
```
Comment (PwzXxm) [inline internal/core/src/exec/operator/FilterNode.cpp:143]: Why this must be met? Is it possible that user gives a small `batch_size`?

File: internal/core/src/exec/operator/FilterNode.cpp [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+            } else {
+                lo = mid + 1;
+            }
+        }
+    }
+    return lo;
+}
+
+RowVectorPtr
+PhyFilterNode::GetOutput() {
```
Comment (PwzXxm) [inline internal/core/src/exec/operator/FilterNode.cpp:75]: Does it support RangeSearch?

**Comment by chasingegg** [comment]
rerun cpp-unit-test

**Comment by chasingegg** [comment]
rerun go-sdk

**Comment by chasingegg** [comment]
rerun cpp-unit-test

**Comment by chasingegg** [comment]
Rename `post filter` to `iterative filter` to indicate the iterative execution model.

**Comment by chasingegg** [comment]
rerun go-sdk

File: internal/core/src/segcore/ChunkedSegmentSealedImpl.cpp [hunk 7]
```diff
@@ -827,6 +831,23 @@ ChunkedSegmentSealedImpl::chunk_view_impl(FieldId field_id,
...
 
+std::pair<std::vector<std::string_view>, FixedVector<bool>>
+ChunkedSegmentSealedImpl::chunk_view_by_offsets(
+    FieldId field_id,
+    int64_t chunk_id,
+    const FixedVector<int32_t>& offsets) const {
+    std::shared_lock lck(mutex_);
+    AssertInfo(get_bit(field_data_ready_bitset_, field_id),
+               "Can't get bitset element at " + std::to_string(field_id.get()));
+    auto& field_meta = schema_->operator[](field_id);
```
Comment (czs007) [inline internal/core/src/segcore/ChunkedSegmentSealedImpl.cpp:842]: unnecessary ?

**Comment by chasingegg** [comment]
rerun go-sdk

**Comment by chasingegg** [comment]
rerun go-sdk

File: internal/proxy/search_util.go [hunk 8]
```diff
@@ -200,6 +205,7 @@ func parseSearchInfo(searchParamsPair []*commonpb.KeyValuePair, schema *schemapb
 			GroupByFieldId:  groupByFieldId,
 			GroupSize:       groupSize,
 			StrictGroupSize: strictGroupSize,
+			Hints:           hints,
```
Comment (czs007) [inline internal/proxy/search_util.go:208]: better to not fill this member if hints equal empty. 
You can change this in next pr .

**Pull request merged**